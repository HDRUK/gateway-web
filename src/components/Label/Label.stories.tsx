import type { Meta, StoryObj } from "@storybook/react";
import Label from "./Label";

/** Mui documentation: https://mui.com/material-ui/react-label */

const meta: Meta<typeof Label> = {
    component: Label,
    tags: ["autodocs"],
};

export default meta;

type Story = StoryObj<typeof Label>;

export const Default: Story = {
    args: {
        label: "Label",
        required: true,
    },
};
