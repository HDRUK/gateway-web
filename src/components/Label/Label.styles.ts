import { css } from "@emotion/react";

export const label = ({ required }: { required: boolean }) =>
    css({
        ...(required && {
            ":after": {
                content: '"*"',
            },
        }),
    });
