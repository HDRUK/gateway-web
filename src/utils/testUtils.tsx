import React, { ReactNode } from "react";
import {
    render,
    renderHook,
    RenderHookOptions,
    RenderOptions,
    RenderResult,
    RenderHookResult,
} from "@testing-library/react";
import { SWRConfig } from "swr";
import { CacheProvider, ThemeProvider } from "@emotion/react";
import ActionBarProvider from "@/providers/ActionBarProvider";
import DialogProvider from "@/providers/DialogProvider";
import theme from "@/config/theme";
import createEmotionCache from "@/config/createEmotionCache";

const clientSideEmotionCache = createEmotionCache();

const Wrapper = ({ children }: { children: ReactNode }) => {
    return (
        <SWRConfig
            value={{
                provider: () => new Map(),
            }}>
            <CacheProvider value={clientSideEmotionCache}>
                <ThemeProvider theme={theme}>
                    <ActionBarProvider>
                        <DialogProvider>{children}</DialogProvider>
                    </ActionBarProvider>
                </ThemeProvider>
            </CacheProvider>
        </SWRConfig>
    );
};

interface OptionProps
    extends RenderOptions<
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement,
        HTMLElement
    > {
    wrapperProps: Record<string, unknown>;
}

const customRender = (
    ui: React.ReactElement<
        unknown,
        string | React.JSXElementConstructor<unknown>
    >,
    options?: OptionProps
): RenderResult => {
    const { wrapperProps, ...rest } = options || {};

    return render(ui, {
        wrapper: props => <Wrapper {...props} {...wrapperProps} />,
        ...rest,
    });
};

interface OptionHookProps
    extends RenderHookOptions<
        unknown,
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement,
        HTMLElement
    > {
    wrapperProps: Record<string, unknown>;
}

const customRenderHook = (
    ui: (initialProps: unknown) => unknown,
    options?: OptionHookProps
): RenderHookResult<unknown, unknown> => {
    const { wrapperProps, ...rest } = options || {};
    return renderHook(ui, {
        wrapper: props => <Wrapper {...props} {...wrapperProps} />,
        ...rest,
    });
};

export * from "@testing-library/react";

export { customRenderHook as renderHook };
export { customRender as render };
