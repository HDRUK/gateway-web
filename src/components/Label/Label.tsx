/** @jsxImportSource @emotion/react */

import { Typography } from "@mui/material";
import { useTheme } from "@emotion/react";
import { ComponentPropsWithoutRef } from "react";
import * as styles from "./Label.styles";

interface LabelProps extends ComponentPropsWithoutRef<"label"> {
    label: string;
    required?: boolean;
}

const Label = ({ label, required }: LabelProps) => {
    const theme = useTheme();

    return (
        <Typography variant="body2" css={styles.label({ theme, required })}>
            {label}
        </Typography>
    );
};

Label.defaultProps = {
    required: false,
};

export default Label;
