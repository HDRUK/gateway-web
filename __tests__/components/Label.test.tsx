import React from "react";
import Label from "@/components/Label";
import { render, screen } from "../testUtils";

describe("Label", () => {
    it("should render label", async () => {
        render(<Label label="This is a label" />);

        expect(screen.getByText("This is a label")).toBeInTheDocument();
    });
    it("should render asterick", async () => {
        const wrapper = render(<Label label="This is a label" required />);

        expect(wrapper.container).toMatchSnapshot();
    });
});
