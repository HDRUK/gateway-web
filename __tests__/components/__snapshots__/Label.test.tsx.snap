// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Head should render asterick 1`] = `
<div>
  <p
    class="MuiTypography-root MuiTypography-body2 mui-style-uahsa5-MuiTypography-root-Label"
  >
    This is a label
  </p>
</div>
`;
